### PR TITLE
Merge and Split Tickets

### PR DESCRIPTION
**Pull request by cconard96**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #914 

Work in progress. Open to any feedback about the implementation.

Done:

- [x]  Merge a ticket and its followups as followups of another ticket. Merged ticket is deleted.
- [x]  "Promote" a followup to a new ticket. A "son" relation is created. New ticket is son of original ticket.

To-do:

- [x]  Test rights
- [x]  Add new event logging for merge and promote actions
- [x]  Fix timeline ordering for merged followups (Don't just reassign items_id?)
- [x]  Fix style issues with new buttons in ticket and followup
- [x]  Add confirmation/warning for merge as ticket currently gets deleted

Should a ticket that gets merged as a followup be deleted or closed?
Should we use "split" or "promote" as the action name for turning a followup into a new ticket?

### CODE REVIEW

**Comment by orthagh** [comment]
Ok first functional review:

### UX is not ok

- "Merge as Followup" should be a massive action and not an additional button in primary form
- "Promote to ticket":
  * it shouldn't be in edit form (i got an hard time to find the feature), i think an icon side of edit one could be ok. (see https://fontawesome.com/v4.7.0/icon/code-fork for iconography)
  * You should open new ticket form in a complete new context: i think about a modal here (or, if it's too complicated, a redirection to new ticket form)

### Tech review

- purge of promoted followups should be done in `Ticket::post_addItem` method

**Comment by trasher** [comment]
Could you please rebase your branch on current master?

There were a lot of changes, and some deprecations (ike using a string for `CommonDBTM::find()`).

File: inc/ticket.class.php [hunk 1]
```diff
@@ -2639,11 +2648,118 @@ function getSpecificMassiveActions($checkitem = null) {
...
+         if (Ticket::canCreate() && Ticket::canDelete()) {
+            $actions[__CLASS__.MassiveAction::CLASS_ACTION_SEPARATOR.'merge_as_followup']
+               = __('Merge as Followup');
+         }
       }
       return $actions;
    }
 
 
+   /**
```
Comment (trasher) [inline inc/ticket.class.php]: drop docblock

File: inc/ticket.class.php [hunk 2]
```diff
@@ -2639,11 +2648,118 @@ function getSpecificMassiveActions($checkitem = null) {
...
+               'confirm'   => 'Confirm the merge? This ticket will be deleted!'
+            ]);
+            echo "</p></div>";
+            return true;
+      }
+      return parent::showMassiveActionsSubForm($ma);
+   }
+
+
+   /**
```
Comment (trasher) [inline inc/ticket.class.php]: drop docblock

File: inc/ticket.class.php [hunk 3]
```diff
@@ -2639,11 +2648,118 @@ function getSpecificMassiveActions($checkitem = null) {
...
+    * @see CommonDBTM::processMassiveActionsForOneItemtype()
+   **/
+   static function processMassiveActionsForOneItemtype(MassiveAction $ma, CommonDBTM $item,
+                                                       array $ids) {
+
+      switch ($ma->getAction()) {
+         case 'merge_as_followup' :
+            $input = $ma->getInput();
+            foreach ($ids as $id) {
+               //$input2 = ['id' => $id];
```
Comment (trasher) [inline inc/ticket.class.php]: dead code

File: inc/ticket.class.php [hunk 4]
```diff
@@ -5090,19 +5222,24 @@ function showForm($ID, $options = []) {
             } else {
                if ($this->canDeleteItem()) {
                   echo "<input type='submit' class='submit' name='delete' value='".
-                         _sx('button', 'Put in trashbin')."'>";
+                         _sx('button', 'Put in trashbin')."'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;";
```
Comment (trasher) [inline inc/ticket.class.php]: Would be better to use some CSS to "make space", no?

File: inc/ticket.class.php [hunk 5]
```diff
@@ -5090,19 +5222,24 @@ function showForm($ID, $options = []) {
...
                }
             }
             echo "<input type='hidden' name='_read_date_mod' value='".$this->getField('date_mod')."'>";
             echo "</div>";
          } else {
             echo "<div class='tab_bg_2 center'>";
-            echo "<input type='submit' name='add' value=\""._sx('button', 'Add')."\" class='submit'>";
+            $add_params = ['name' => 'add'];
+            if ($options['_promoted_fup_id']) {
+               $add_params['confirm'] = 'Confirm the promotion? This followup will be deleted!';
```
Comment (trasher) [inline inc/ticket.class.php]: Sould not this been translatable?

File: inc/ticket.class.php [hunk 6]
```diff
@@ -5090,19 +5222,24 @@ function showForm($ID, $options = []) {
...
+            }
+            echo Html::submit(_x('button', 'Add'), $add_params);
             if ($tt->isField('id') && ($tt->fields['id'] > 0)) {
                echo "<input type='hidden' name='_tickettemplates_id' value='".$tt->fields['id']."'>";
                echo "<input type='hidden' name='_predefined_fields'
                       value=\"".Toolbox::prepareArrayForInput($predefined_fields)."\">";
             }
+            echo "<input type='hidden' name='_promoted_fup_id' value='".$options['_promoted_fup_id']."'>";
```
Comment (trasher) [inline inc/ticket.class.php]: Maybe better to use a `Html` method

**Comment by cconard96** [comment]
I've rebased and made requested changes. I will do more work on it later tonight. I am not really happy with how followups are created during a merge. They currently get recreated and end up with the same creation date. May be a good use of #4938 and we should maybe mark the new followups as being merged from another ticket.

Thoughts?

**Comment by trasher** [comment]
>  we should maybe mark the new followups as being merged from another ticket

I entirely agree :)

**Comment by trasher** [comment]
Thank you @all :)

**Pull request merged**